// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.



// Compile using:
// python3 -m grpc_tools.protoc -I./ --python_out=. --pyi_out=. --grpc_python_out=. ./chat.proto

syntax = "proto3";

package helloworld;


// The chat service definition.
service ChatService {
  rpc Login (User) returns (User) {}
  rpc Logout (User) returns (User) {}
  rpc Register (User) returns (User) {}
  rpc GetUsers (SearchQuery) returns (ListOfUsernames) {}
  rpc Connect (ChatMessage) returns (User) {} // Connect and disconnect take in ChatMessage as request type because that has sender and recipient fields
  rpc Disconnect (User) returns (User) {}
  rpc ChatStream (User) returns (stream ChatMessage) {} // Listening stream for client
  rpc SendMessage (ChatMessage) returns (Empty) {}
  rpc Delete (User) returns (Empty) {}
}


message Empty {}

message Request {
  string message = 1;
}

message RequestResponse {
  bool success = 1;
}

message SearchQuery {
  string query = 1;
}

message ListOfUsernames {
  repeated string usernames = 1;
}

message ChatMessage {
  string message = 1;
  string to = 2;
  string sender = 3;
}

message User {
  string name = 1;
}


